prometheus/wrap.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// WrapWith returns a Registerer wrapping the provided Registerer. Collectors
+// registered with the returned Registerer will be registered with the wrapped
+// Registerer in a modified way. The modified Collector adds the provided Labels
+// to all Metrics it collects (as ConstLabels). The Metrics collected by the
+// unmodified Collector must not have any of those labels on their own.
+//
+// WrapWith provides a way to add fixed labels to a subset of Collectors. It
+// should not be used to add fixed labels to all metrics exposed.
+//
+// The Collector example demonstrates a use of WrapWith.
+func WrapWith(labels Labels, reg Registerer) Registerer {
```

Comment:
I don't think the "Wrap" terminology is necessary and the "wrapped" thing should probably be the first argument, then the labels can be variadic which is a more natural interface I think. I'd suggest to rename the function to `WithLabels`.

Reply:
I mostly agree with these comments, although it's *slightly* weird to have `WithXXX()` functions on the package level rather than on a concrete type. Since this already returns us a non-global registry anyway, we might as well define the method on a registry itself and then people can call it on the default registry if they want? `myRegistry.WithLabels(...)`. Like contexts do it.

Reply:
(ok contexts were a stupid example, they define package-level functions instead of methods, but that is because the context type is the only one that's relevant in that whole package)

Reply:
I agree it would make sense to have them just on the registry itself, but throws up the question whether they should be part of the interface (which seems odd) or part of the registry types, which would cause type switches/casts to wrap already wrapped registries. It seems to me that the package level functions are actually a reasonable tradeoff.